server/common/diffexpdu/postingslist.py

```diff
@@ -0,0 +1,426 @@
+    """
+    Return the length of the deflated posting list encoded in buf.
+
+    Returns
+    -------
+    int
+        Length of deflated posting list in bytes.
+    """
+    buf = memoryview(buf)
+    if offset > 0:
+        buf = buf[offset:]
+
+    _, block_descriptions = _decode_header(buf)
+    n_blocks = len(block_descriptions)
+    return (
+        headerPacker.size
+        + n_blocks * _BlockDescription.packed_length
+        + sum([block.n_bytes for block in block_descriptions])
+    )
+
+
+def deflate_postings_lists(disjoint_postings_lists: DisjointPostingsLists, sorted: bool = False) -> bytearray:
+    """
+    Efficiently one or more (max 8) monotonically increasing list of unsigned integers into a byte stream.
+    Lists must be disjoint (the current implementation may not fail if lists are not unique, but future
+    changes may enforce/require this)
+    """
+    assert 0 < len(disjoint_postings_lists) <= 8
+
+    def normalize_plist(pl):
+        if type(pl) is list:
+            pl = np.array(pl, dtype=np.uint32)
+        if pl.dtype != np.uint32:
+            pl = pl.astype(np.uint32)
+        if not sorted:
+            pl.sort()
+        return pl
+
+    disjoint_lists = tuple((normalize_plist(plist) for plist in disjoint_postings_lists))
+    all_partitions = partition_lists(disjoint_lists)
+    all_partitions.sort(key=lambda partition: partition.key)
+
+    encoded_blocks = []
+    for _, parts in itertools.groupby(all_partitions, key=lambda partition: partition.key):
+        # encode each list as individual blocks.
+        parts_tuple = tuple(parts)
+        assert len(parts_tuple) <= 8
+        encoded_blocks.extend(_encode_partition(part) for part in parts_tuple)
+
+    block_count = len(encoded_blocks)
+    encoded_postings = bytearray(4)
+    headerPacker.pack_into(encoded_postings, 0, MAGIC_NUMBER, 0, block_count - 1)
+
+    # encode block descriptions
+    for hdr, values in encoded_blocks:
+        assert hdr.n_bytes == len(values)
+        encoded_postings.extend(hdr.to_bytes())
+
+    assert len(encoded_postings) == headerPacker.size + blockDescriptionPacker.size * block_count
+
+    # encode blocks
+    for _, values in encoded_blocks:
+        encoded_postings.extend(values)
+
+    return encoded_postings
+
+
+def _choose_block_type(part: _ListPartition) -> _BlockDescription.BlockType:
+    """
+    Given a partition, choose a block encoding.  Heuristics live here.
+    """
+    n_elem = part.end_idx - part.start_idx
+    assert 0 < n_elem <= 2 ** 16
+    assert (part.arr[part.start_idx] >> 16) == (part.arr[part.end_idx - 1] >> 16)
+
+    interval = (part.arr[part.end_idx - 1] & 0xFFFF) + 1 - (part.arr[part.start_idx] & 0xFFFF)
+    # inv_n_elem = interval - n_elem
+    density = n_elem / interval
+
+    minBitArrayThreshold = 2 ** 11
+    maxBitArrayThreshold = 2 ** 16 - minBitArrayThreshold
+    if (minBitArrayThreshold < n_elem < maxBitArrayThreshold) and (0.125 < density < 0.875):
```

Comment:
I'm guessing this is to do with byte/delta shuffling which reduces the size by ~50%? If so what about starting with that uint encoding and falling back on bitarray if the reduction is not realized, i.e. the block is > 2^16 bits?

Reply:
It primarily has to do with the subsequent deflate compression performance (deflate is applied to all block types).  The amount of compression is highly dependent on the actual list values, not their encoding.  While it is super easy to predict the cut-off for uncompressed blocks, it is not easy to do it for compressed blocks (without calling deflate()).

And the reason I didn't want to call deflate() is that it is expensive - doing the prospective compression in all formats, and then picking the smallest, would at least double the cost of encoding a block.  For large cell sets, this gets prohibitive on the front-end.

So, what I did is generate a very large set of test cases (over 75K) from real data, and empirically determine a reasonable cut-off.  Turns out that the heuristic appears reasonable from first principles (eg, small sets are better to encode with a discrete list,  very large lists are better to code with an inverted list, and all others are better as a bitarray).  The density clamp turns out to control for a very rough estimate on entropy, further refining the heuristic guess of how well deflate will work.

I'm happy to share the code I used to derive these specific values if you are interested.  I didn't think it was worth putting in the repo as it is not a runtime dependency (but LMK if you disagree).


As an aside -- the reason more exotic block types (eg, multi-list disjoint encodings, etc) did not provide significant benefits is because we use deflate on all block types.

Reply:
Thanks for this context. This all makes sense. I think it would be helpful to have a version of this explanation in the code if possible, to aid future readers.

Reply:
both the Py and TS code has the same heuristic, so I'll add a comment there pointing to the dev_docs, and add a section there describing the tradeoffs/heuristic.